File: LeetCode/199.Binary_Tree_Right_Side_View/Mario/199.Binary_Tree_Right_Side_View.py
#!/usr/bin/env python3


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        res = []

        if not root:
            return res

        level = [root]

        while level:
            rightMostValue = 0
            array = []
            for node in level:
                rightMostValue = node.val
                if node.left:
                    array.append(node.left)
                if node.right:
                    array.append(node.right)
            res.append(rightMostValue)
            level = array

        return res


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        res = []

        def dfs(node, depth):
            if not node:
                return

            if depth == len(res):
                res.append(node.val)

            # right first
            dfs(node.right, depth + 1)
            dfs(node.left, depth + 1)

        dfs(root, 0)

        return res

File: LeetCode/199.Binary_Tree_Right_Side_View/199.Binary_Tree_Right_Side_View.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func rightSideView(_ root: TreeNode?) -> [Int] {
        guard let root = root else { return [] }

        var q = Deque<TreeNode>()
        q.append(root)

        var res: [Int] = []
        while !q.isEmpty {
            let size = q.count

            for i in 0 ..< size {
                let node = q.popFirst()!
                if i == size - 1 {
                    res.append(node.val)
                }

                if let left = node.left {
                    q.append(left)
                }
                if let right = node.right {
                    q.append(right)
                }
            }
        }

        return res
    }
}
